File: domain/meaning_sources_cleaning.py
"""Gloss cleaning utilities for meaning sources."""

from __future__ import annotations

from typing import Sequence, List

GlossList = List[str]


def _clean_list(xs: object) -> GlossList:
    if not isinstance(xs, list):
        return []
    out: GlossList = []
    for x in xs:
        if isinstance(x, str):
            s = x.strip()
            if s:
                out.append(s)
    return out


def _best_effort_clean_glosses(glosses: Sequence[str] | object) -> GlossList:
    """Best-effort gloss cleaning suitable for UI labels/tooltips."""
    if isinstance(glosses, (list, tuple)):
        return _clean_list(list(glosses))
    return []


def clean_glosses_for_display(glosses: Sequence[str] | object) -> GlossList:
    """Public cleaning helper for UI labels/tooltips."""
    return _best_effort_clean_glosses(glosses)

File: flutter_app/lib/domain/meaning_sources_cleaning.dart
List<String> _cleanList(dynamic xs) {
  if (xs is! List) {
    return [];
  }
  final out = <String>[];
  for (final x in xs) {
    if (x is String) {
      final s = x.trim();
      if (s.isNotEmpty) {
        out.add(s);
      }
    }
  }
  return out;
}

List<String> cleanGlossesForDisplay(dynamic glosses) {
  if (glosses is List) {
    return _cleanList(glosses);
  }
  return [];
}
